### DISCUSSION TITLE
rdDepictor.GenerateDepictionMatching2DStructure not working with rdDepictor.SetPreferCoordGen(True)

### DISCUSSION THREAD

**Discussion by DavidACosgrove** [rdkit/rdkit#4994] [Q&A]
Hi,

I'm using 2021.09.4 installed via Conda on macOS BigSur 11.6.3.

I'm trying to use GenerateDepictionMatching2DStructure to fix part of a macrocycle to a preferred layout taken from a reference SDF.  It works ok with rdDepictor.SetPreferCoord(False) i.e. using the RDKit 2D coords generator (and also passing forceRDKit=True to GenerateDepictionMatching2DStructure) but this lays out the rest of the macrocycle in a somewhat suboptimal way.  CoordGen does a much better layout, but I can't persuade it to pick up the constraint coordinates.  Is there a way of using GenerateDepictionMatching2DStructure with CoordGen?  

As a simple test case, python -m unittest testDepictor.TestCase.testGenerate2DDepictionRefPatternAtomMap in $RDBASE/Code/GraphMol/Depictor/Wrap fails if you put rdDepictor.SetPreferCoord(True) at the top of the test.

Thanks,
Dave

**Comment by ptosco** [comment]
@DavidACosgrove `GenerateDepictionMatching2DStructure` works with `CoordGen`, but `CoordGen` takes a different approach at constrained depiction: it will first draw the individual bits, and then connect them to match the reference. However, in that specific case the pyrazole in the reference is drawn differently from the way `CoordGen` draws a pyrazole. So even though `CoordGen` will RMS fit its layout to the reference, the layout looks a bit different from the reference, as the pyrazole was first redrawn, and then RMS-fitted.
Instead, RDKit freezes the coordinates of the whole reference and builds the rest of the molecule around it, so the core is indeed identical to the reference.
You can see that constrained depiction works also with `CoordGen` by changing the test case a bit, as I did below:
```
from rdkit import Chem
from rdkit.Chem import rdDepictor

rdDepictor.SetPreferCoordGen(True)

indazoleMolblock = """
     RDKit          2D

  9 10  0  0  0  0  0  0  0  0999 V2000
   -6.0878    2.4335    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -7.3867    1.6835    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -7.3867    0.1833    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -6.0878   -0.5666    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -4.7887    0.1833    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -4.7887    1.6835    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.4897   -0.5664    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1906    1.6833    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.1906    0.1835    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  2  0
  2  3  1  0
  3  4  2  0
  4  5  1  0
  5  6  2  0
  6  1  1  0
  8  9  2  0
  6  8  1  0
  7  9  1  0
  7  5  1  0
M  END"""
indazoleRef = Chem.MolFromMolBlock(indazoleMolblock)
cycloheptylPyrazole = Chem.MolFromSmiles("c1cc(C2CCCCCC2)[nH]n1")

# test using refPattern
refPatt = Chem.MolFromSmarts("a1aan[nH]1")
rdDepictor.GenerateDepictionMatching2DStructure(cycloheptylPyrazole, indazoleRef, refPatt=refPatt)
((4, 2), (5, 1), (7, 0), (8, 11), (6, 10))
cycloheptylPyrazole
```
![image](https://user-images.githubusercontent.com/5244385/152825421-623b6216-69ae-440a-9252-3ea6e7e38015.png)
If I do a standard layout of a substituted cycloheptylpyrazole, this is what I get:
```
cycloheptylPyrazoleSubstituted = Chem.MolFromSmiles("c1c(CCOCCOCCO)c(C2CCC(NCCNCCNCC)CCC2)[nH]n1")
cycloheptylPyrazoleSubstituted
```
![image](https://user-images.githubusercontent.com/5244385/152825774-792abb7d-787b-4674-9d7e-cb0a97d4407d.png)
If I do a constrained layout, the core is indeed oriented as in the reference, as I would expect:
```
rdDepictor.GenerateDepictionMatching2DStructure(cycloheptylPyrazoleSubstituted, cycloheptylPyrazole)
((0, 0),
 (1, 1),
 (2, 11),
 (3, 12),
 (4, 13),
 (5, 14),
 (6, 15),
 (7, 25),
 (8, 26),
 (9, 27),
 (10, 28),
 (11, 29))
cycloheptylPyrazoleSubstituted
```
![image](https://user-images.githubusercontent.com/5244385/152826150-6539d4a1-890b-4fb5-b496-e9ba025b7254.png)

**Reply by DavidACosgrove** [reply]
Thanks for the prompt reply, @ptosco.  These are the examples I have problems with.  They are rapamycin and a simple derivative.  The highlights are the constrained bits so pretty much everything except a methyl (atom 44).  The images are rapamycin itself, the result from CoordGen and the result from RDKit.  I think you'd agree that the CoordGen result is a long way from what one might have hoped for.
<img width="476" alt="image" src="https://user-images.githubusercontent.com/9198870/152832558-9ba21b61-16c6-41b7-80e5-5de6d0689073.png">
<img width="479" alt="image" src="https://user-images.githubusercontent.com/9198870/152832698-30d51f78-a64b-4225-8e0f-81848e1b94ea.png">
<img width="477" alt="image" src="https://user-images.githubusercontent.com/9198870/152832780-5acee08f-506f-4367-b290-b732d874e3d5.png">

**Comment by ptosco** [comment]
@DavidACosgrove I do agree this example is disappointing, It's been a while since I have last looked at the `CoordGen` constrained depiction code, but from my memories it really works in a very different way compared to the RDKit one. You might want to report this example on [Schrodinger's `CoordGen` bug tracker](https://github.com/schrodinger/coordgenlibs/issues).
